### PR TITLE
mxnet io 

### PR DESCRIPTION
**Pull request by yanqingmen**
add io related classes and jni funcs;
modified linux setting

### CODE REVIEW

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    */
+  def getIndex(): List[Long]
+
+}
+
+/**
+  * DataIter built in MXNet.
+  * @param handle the handle to the underlying C++ Data Iterator
+  */
+class MXDataIter(var handle: DataIterHandle) extends DataIter {
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: Could 'handle' be val ?

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc [hunk 2]
```diff
@@ -489,3 +489,179 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxNDArrayFree(JNIEnv * env, jo
...
+    keys[i] = (char*)env->GetStringUTFChars(jkey, 0);
+    jval = (jstring) env->GetObjectArrayElement(jvals, i);
+    vals[i] = (char*)env->GetStringUTFChars(jval, 0);
+  }
+
+//  printf("paramSize: %d\n", paramSize);
+//  for(int i=0; i<paramSize; i++) {
+//    printf("key: %s\t",keys[i]);
+//    printf("value: %s\n", vals[i]);
+//  }
```
Comment (javelinjs) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: remove debug codes please

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc [hunk 3]
```diff
@@ -489,3 +489,179 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxNDArrayFree(JNIEnv * env, jo
...
+  jclass hClass = env->GetObjectClass(dataIterHandle);
+  jfieldID ptr = env->GetFieldID(hClass, "value", "J");
+  env->SetLongField(dataIterHandle, ptr, (long)out);
+
+  //release const char*
+  for(int i=0; i<paramSize; i++) {
+      jkey = (jstring) env->GetObjectArrayElement(jkeys, i);
+      env->ReleaseStringUTFChars(jkey,(const char*)keys[i]);
+      jval = (jstring) env->GetObjectArrayElement(jvals, i);
+      env->ReleaseStringUTFChars(jval,(const char*)vals[i]);
```
Comment (javelinjs) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: In my understanding, ReleaseStringUTFChars can only release char* which generated by GetStringUTFChars, but not 'new' in c++. Could you make sure it won't cause memory leak here?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala (hunk 1 above)
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: and remember to add finalize() to this class

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc [hunk 4]
```diff
@@ -489,3 +489,179 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxNDArrayFree(JNIEnv * env, jo
...
+  //set field
+  setLongField(env, outSize, (long)coutSize);
+  // scala.collection.mutable.ListBuffer append method
+  jclass listClass = env->FindClass("scala/collection/mutable/ListBuffer");
+  jmethodID listAppend = env->GetMethodID(listClass,
+    "$plus$eq", "(Ljava/lang/Object;)Lscala/collection/mutable/ListBuffer;");
+
+//  printf("outSize: %ld\n", coutSize);
+  for(int i=0; i<coutSize; i++) {
+//    printf("%ld\t", coutIndex[i]);
```
Comment (javelinjs) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: remember to remove temporary debug messages

File: scala-package/core/src/test/scala/ml/dmlc/mxnet/IOSuite.scala [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      "shuffle" -> "1",
+      "flat" -> "1",
+      "silent" -> "0",
+      "seed" -> "10"
+    )
+
+    //println("create MNISTIter")
+    val mnist_iter = IO.createIterator("MNISTIter", params)
+    //test_loop
+    mnist_iter.reset()
```
Comment (javelinjs) [inline scala-package/core/src/test/scala/ml/dmlc/mxnet/IOSuite.scala]: mnist_iter -> mnistIter
remove println or just uncomment it

**Comment by tqchen** [comment]
On comment on exposing the modules. You can consider use of reflection, which automatically creates functions under certain scope, this is used to create Operator functions in python, R and julia version of library.

**Comment by yanqingmen** [comment]
yeah, reflection would be a more natural way. I'll try to figure out this

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+
+/**
+  * class batch of data
+  * @param data
+  * @param label
+  * @param index
+  * @param pad
+  */
+class DataBatch(val data: NDArray,
```
Comment (terrytangyuan) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: Consider using case class for many benefits?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    * @param handle
+    * @param params
+    * @return
+    */
+  private def creator(handle: DataIterCreator)(
+              params: Map[String, String]): DataIter = {
+    val out = new DataIterHandle
+    val keys = params.keys.toArray
+    val vals = params.values.toArray
+    checkCall(_LIB.mxDateIterCreateIter(handle, keys, vals, out))
```
Comment (terrytangyuan) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: This should be mxDataIterCreateIter.

File: scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc [hunk 8]
```diff
@@ -489,3 +489,180 @@ JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxNDArrayFree(JNIEnv * env, jo
...
+    int ret = MXListDataIters(&outSize, &outArray);
+    for (int i = 0; i < outSize; ++i) {
+      DataIterCreator chAddr = outArray[i];
+      jobject chObj = env->NewObject(chClass, chConstructor, (long)chAddr);
+      env->CallObjectMethod(creators, listAppend, chObj);
+    }
+    return ret;
+}
+
+JNIEXPORT jint JNICALL Java_ml_dmlc_mxnet_LibInfo_mxDateIterCreateIter
```
Comment (terrytangyuan) [inline scala-package/native/src/main/native/ml_dmlc_mxnet_native_c_api.cc]: mxDataIterCreateIter instead

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  /**
+    * Get label of current batch
+    * @return the label of current batch
+    */
+  override def getLabel(): NDArray = {
+    val out = new NDArrayHandle
+    checkCall(_LIB.mxDataIterGetLabel(handle, out))
+    return new NDArray(out)
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/IO.scala]: should new NDArray(out, writable=false), default writable is true
same as getData()

**Comment by javelinjs** [comment]
@yanqingmen It seems that you have left some methods unimplemented, e.g. next() in MXDataIter, remember to implement it later.

And you happened to add scala-package/scala_test_results.txt to git by mistake.

**Comment by yanqingmen** [comment]
@javelinjs yes, further work is still needed, only basic function is implemented now. I'll finish this later.

**Comment by javelinjs** [comment]
@tqchen Thank you for pointing out. I have noticed the way used to build Operator and Symbol, we'll follow the approaches. And we did have taken advantages of reflection in NDArray.scala, while I tend to make it a little bit more strong-typed, which could be more friendly to users.

**Comment by javelinjs** [comment]
LGTM so far

**Pull request merged**